File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if n > 0 and (head is None or head.next is None):
            return None
        size = 0
        cur = head
        while cur:
            cur = cur.next
            size += 1
        steps = size - n
        cur = head
        if steps == 0:
            return head.next
        for i in range(steps - 1):
            cur = cur.next
        temp = cur
        cur = cur.next.next
        temp.next = cur
        return head

File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode last = head;
        if(last == null){
            return null;
        }
        ListNode cur = head.next;
        if(cur == null){
            return head;
        }
        last.next = null;
        while (cur != null){
            ListNode temp = cur.next;
            cur.next = last;
            last = cur;
            cur = temp;
        }
        return last;
    }
}
